services/github_service.py

```diff
@@ -0,0 +1,126 @@
+                exception, RateLimitExceededException) else rate_limits.graphql
+
+            reset_timestamp = timegm(rate_limit_to_use.reset.timetuple())
+            now_timestamp = timegm(gmtime())
+            time_until_core_api_rate_limit_resets = (
+                reset_timestamp - now_timestamp) if reset_timestamp > now_timestamp else 0
+
+            wait_time_buffer = 5
+            sleep(time_until_core_api_rate_limit_resets +
+                  wait_time_buffer if time_until_core_api_rate_limit_resets else 0)
+            return func(*args, **kwargs)
+
+    return decorator
+
+
+class GithubService:
+    def __init__(
+        self,
+        org_token: str,
+        org_name: str = "ministryofjustice"
+    ) -> None:
+        self.github_client_core_api = Github(org_token)
+        self.org_name = org_name
+        self.github_client_gql_api: Client = Client(transport=AIOHTTPTransport(
+            url="https://api.github.com/graphql",
+            headers={"Authorization": f"Bearer {org_token}"},
+        ), execute_timeout=120)
+        self.github_client_rest_api = Session()
+        self.github_client_rest_api.headers.update(
+            {
+                "Accept": "application/vnd.github+json",
+                "Authorization": f"Bearer {org_token}",
+            }
+        )
+
+    @retry(stop_max_attempt_number=3, wait_fixed=2000)
+    @retries_github_rate_limit_exception_at_next_reset_once
+    def get_all_private_non_archived_repos(self) -> list[github.Repository]:
+
+        org = self.github_client_core_api.get_organization(self.org_name)
+        non_archived_private_repos = [repo for repo in org.get_repos(type="private") if not repo.archived]
+
+        return non_archived_private_repos
+
+    @retry(stop_max_attempt_number=3, wait_fixed=2000)
+    @retries_github_rate_limit_exception_at_next_reset_once
+    def get_all_internal_non_archived_repos(self) -> github.PaginatedList:
+
+        org = self.github_client_core_api.get_organization(self.org_name)
+        non_archived_internal_repos = [repo for repo in org.get_repos(type="internal") if not repo.archived]
+
+        return non_archived_internal_repos
+
+    @retry(stop_max_attempt_number=3, wait_fixed=2000)
+    @retries_github_rate_limit_exception_at_next_reset_once
+    def get_workflows_per_repo(self, repo: github.Repository) -> github.PaginatedList:
+
+        workflows = repo.get_workflows()
+
+        return workflows
+
+    @retry(stop_max_attempt_number=3, wait_fixed=2000)
+    @retries_github_rate_limit_exception_at_next_reset_once
+    def get_workflow_runs(self, workflow: github.Workflow) -> github.PaginatedList: 
+
+        workflow_runs = workflow.get_runs()
+
+        return workflow_runs
+
+    @retry(stop_max_attempt_number=3, wait_fixed=2000)
+    @retries_github_rate_limit_exception_at_next_reset_once
+    def get_workflow_runs_repo(self, repo: github.Repository, created: str) -> github.PaginatedList: 
```

Comment:
Name is a bit confusing, think "get_workflow_runs_per_repo" would be better

Reply:
Fixed.